### ISSUE TITLE
K3po remove need for k3po.start() and k3po.finish()

### ISSUE THREAD

**Issue by dpwspoon** [k3po/k3po#267]

**Comment by dpwspoon** [comment]
### Backwards compatibility proposal, assuming we go with the design

deprecate all old k3po junit methods, and point to newer methods.  Internally have the  deprecated `k3po.start` and `k3po.finish` send a `notify("BOUND")` command.

Note: this does not necessarily solve for k3po scripts that live in other projects, as those scripts will now fire their client channels prior to the @Test method

**Comment by jfallows** [comment]
I like where this is going, @dpwspoon.

Do we need to have implicit `connect` barriers such as `~START` so that `k3po.start()` can be implemented as `k3po.notify("~START")` internally?

```
accept tcp://localhost:8080
connect await ~START
        tcp://localhost:8080
```

Do we also need to require that every `accept` has completed the bind before returning `PREPARED`? If so, how does this evolve when adding `accept await BARRIER` syntax which is the natural next step?

**Comment by jitsni** [comment]
@dpwspoon In some cases, junit cannot send notify("BOUND") since it may be internal. For e.g tcp client and server go through two gateways(which are connected via reverse/forward connectivity) and we don't know when a gateway is BOUND.

**Comment by cmebarrow** [comment]
Whatever we do I would strongly advocate that it should be compatible with our existing spec tests and their consumption in gateway.server, etc. I would hate us to have to change all of that. John's idea of implicit connect barrier (~START) which the (deprecated) start() would notify should hopefully achieve this. Is that what you had in mind John @jfallows?

**Comment by dpwspoon** [comment]
How do we make the injected ~START compatible with the new API.  Perhaps we only inject it if there is not already an await barrier on the connect?  That would make scripts that use the new Junit API break against scripts that don't have an explicit barrier on the connect.

> Do we also need to require that every accept has completed the bind before returning PREPARED? If so, how does this evolve when adding accept await BARRIER syntax which is the natural next step?

Is there a use case for delaying the bind / LISTEN, or only the connect?  i.e. could it be 
```
accept tcp://localhost:8000
connected await 
```

Assuming there is a case where you need an await barrier on the accept, then it seems to me that we could have logic to return the STARTED when all accepts without await barriers have bound... thoughts?

**Comment by jfallows** [comment]
@cmebarrow: yes, with the behavior of the call to `k3po.start()` still implicit, such that `~START` barrier could be notified implicitly. In cases where every `connect` has a custom `BARRIER` then this notifiying of `~START` would be harmless.

Btw, @dpwspoon was discussing this recently too, so it's his idea as well. :wink: 

@dpwspoon: yes, i think there is a use case for delaying the bind, with the following syntax.

```
accept await BARRIER
       tcp://localhost:8080
```

Once `accepted`, then both `read await` and `write await` can be used from within the stream directly as usual.

**Comment by dpwspoon** [comment]
@jfallows, I think we are still missing something in the design and I just reasoned it through with Chris.  Here is an example of blindly injecting the ~START that breaks if the script is standalone (not written in a spec project and thus doesn't need a barrier to pass).  And it using the new API.

```
connect tcp://localhost:8000
connected
write "hello"
read "world"
```

Pseudocode for test code
```java

tcpServerRule.around(k3poRule)

@Test
public void test(){
     tcpServerRule.read("hello")
     tcpServerRule.write("world")
}

```

However, it occurred to us if we knew the Junit API that the test was using at runtime then we could decide whether the injected ~start was needed or not.  There are probably many ways to solve this (Package, class rename)...

**Comment by jfallows** [comment]
@dpwspoon: i was thinking we would keep the `k3po.finish()` and the implicit `k3po.start()` would notify the `~START` barrier.
